### ISSUE TITLE
EmailField doesn't perform server-side validation

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow-components#2984]
### Description

Typing "foo" into the EmailField will cause the value of the EmailField to be "foo" on the server-side, even though the value is invalid.

### Expected outcome

The server-side EmailField should retain its previous value. Similarly when you type in "foo" into BigDecimalField, the server-side field won't accept the value and will retain the previous value.

### Actual outcome

The server-side EmailField.getValue() returns "foo".

### Minimal reproducible example

Paste the following code into Skeleton Starter:

```
@Route("")
public class MainView extends VerticalLayout {

    public MainView() {
        final EmailField emailField = new EmailField("E-mail");
        emailField.addValueChangeListener(e -> System.out.println(emailField.getValue()));
        add(emailField);
    }
}
```

### Steps to reproduce

1. Run the skeleton starter
2. Type in "foo"
3. The component turns red, but the value is transferred server-side and available as a value of the component.

### Environment

- Version of the component / Vaadin platform. Vaadin 14.8.7
- OS. Ubuntu Linux 21.10

**Comment by TatuLund** [comment]
I think you should have `emailField.setPreventInvalidInput(true)` 

In case of `EmailField` how ever it would feel natural to have `true` as default value.

```
@Route("")
public class MainView extends VerticalLayout {

    public MainView() {
        final EmailField emailField = new EmailField("E-mail");
        emailField.setPreventInvalidInput(true);
        emailField.addValueChangeListener(e -> System.out.println(emailField.getValue()));
        add(emailField);
    }
}
```

**Comment by mvysny** [comment]
Uh-oh, `setPreventInvalidInput(true);` makes things far worse: now it's not possible to type in anything into the EmailField.

**Comment by TatuLund** [comment]
Yes, but this seems to work, i.e. you can use field validity status to filter 

```
@Route("email")
public class EmailView extends VerticalLayout {
    
    public EmailView() {
        final EmailField emailField = new EmailField("E-mail");
        emailField.addValueChangeListener(e -> {
            if (!emailField.isInvalid()) {
                Notification.show(e.getValue());
            }
        });
        add(emailField);
    }
}
```

**Comment by jouni** [comment]
> I think you should have `emailField.setPreventInvalidInput(true)`

I would not recommend that. It is not a replacement for working validation features.

**Comment by vursen** [comment]
This issue looks similar to https://github.com/vaadin/flow-components/issues/2985.

> The current behavior is intended. The server-side value is synced on every user input change _except when it cannot be cast to the server-side value type_ which is a limitation coming from using a strongly typed language. For example, an arbitrary `foobar` input string cannot be transformed into a LocalDate in the case of DatePicker but it is acceptable for EmailField whose value is of String type even though that string is not a valid email.
>
> ...
>
> We could consider changing the current behavior for all the field components in the future if the suggested behavior would be proven to cover the already existing use-cases.

For more details, please check out https://github.com/vaadin/flow-components/issues/2985#issuecomment-1268500400.
